mypy/checker.py

```diff
@@ -5572,6 +5569,36 @@ def visit_match_stmt(self, s: MatchStmt) -> None:
             with self.binder.frame_context(can_skip=False, fall_through=2):
                 pass
 
+    def _make_named_statement_for_match(self, s: MatchStmt) -> Expression:
+        """Construct a fake NameExpr for inference if a match clause is complex."""
+        subject = s.subject
+        expressions_to_preserve = (
```

Comment:
IIUC match statement handling re-uses binder in some code paths. Then I I think instead of this list we should do the same that binder does:
```python
from mypy.literals import literal  # name is confusing, but this is what you want
if not literal(expr):
    <make dummy name>
```

Reply:
Nice idea, but are you sure?

https://github.com/python/mypy/blob/fc991a0ac83929d2d250a20b8283ba1229a15e75/mypy/literals.py#L117-L124

This will break `testMatchOperations` case added here, for example, as `1 < 2` will return `LITERAL_YES`, and we do not store types of comparison exprs in binder AFAIC.

Reply:
Right, binder also does `isinstance(expr, (IndexExpr, MemberExpr, NameExpr))` before checking for `literal()`. But then this list will not be simplified much. Even then probably it would make sense to somehow re-use binder logic here.

Btw I am not super-confident about `AssignmentExpr`, what about something like `x := y[z]` (where `z` is _not_ a literal)? This will have troubles with binder.

Reply:
Maybe we can add a common ancestor to `{Str,Bytes,Int,Float,Complex}Expr`? Semantically they should have a "primitive literal" superclass, then this list will look much saner - just "name or primitive".